### ISSUE TITLE
MODIFYCLUSTERSETTING does not work as expected 22.2+

### ISSUE THREAD

**Issue by data-matt** [cockroachdb/cockroach#96172]
**Describe the problem**

When providing the system privilege MODIFYCLUSTERSETTING in 22.1, this setting provided access to alter sql.defaults.* cluster settings.

From 22.2 MODIFYCLUSTERSETTING allows access to all cluster settings.

**To Reproduce**

```
root@localhost:26257/defaultdb> alter role test3001 with modifyclustersetting;
ALTER ROLE

 ./cockroach sql --certs-dir=certs --user test3001


test3001@localhost:26257/defaultdb> SET CLUSTER SETTING diagnostics.reporting.enabled = false;
SET CLUSTER SETTING

````

**Expected behavior**

The user "test3001" shouldn't be able to set cluster settings outside of sql.defaults.

Seen in previous versions: 

```
root@localhost:26257/defaultdb> alter role test3001 with modifyclustersetting;
ALTER ROLE

 ./cockroach sql --certs-dir=certs --user test3001

test3001@:26257/defaultdb> SET CLUSTER SETTING diagnostics.reporting.enabled = false;
warning: cannot save command-line history: write_history: illegal byte sequence
ERROR: only users with the admin role are allowed to set cluster setting 'diagnostics.reporting.enabled'
SQLSTATE: 42501
```

**Additional context**
This violates the principle of least privilege. Once upgraded to 22.2, there will be users that have access to all cluster settings and therefore able to alter security settings such as the hba.conf. 

Ultimately meaning, that in this state we have to revoke access to MODIFYCLUSTERSETTING and users will no longer be able to edit sql.defaults without adjusting roles..

 [ALTER ROLE ALL SET {sessionvar} = {val}](https://www.cockroachlabs.com/docs/v22.2/alter-role#set-default-session-variable-values-for-all-users)

Jira issue: CRDB-23981

**Comment by data-matt** [comment]
The workaround using ALTER ROLE is most likely unsuitable since it requires CREATEROLE privilege.

I thought that maybe users could alter their own role.

```
test3001@:26257/defaultdb> ALTER ROLE test3001 SET datestyle=iso;
ERROR: user test3001 does not have CREATEROLE privilege
SQLSTATE: 42501
```

**Comment by Santamaura** [comment]
@rafiss this change was made as part of the rollout of system privileges. I'm unsure about what the ideal solution is here. We could have it so that only the system privilege allows updates to all settings though this makes the behavior of a system privilege not the same across all situations (now there's will be situation where the system privilege doesn't fall back on the regular privileges) or have both only able to change `sql.defaults.*` but then the system privilege isn't really a global privilege anymore. Any suggestions?

**Comment by rafiss** [comment]
@Santamaura yeah, I recall that we discussed it in this Slack thread: https://cockroachlabs.slack.com/archives/C03BTH2B1ND/p1664819314445229?thread_ts=1664818370.356759&cid=C03BTH2B1ND

It seems like the decision to enlarge the scope of the MODIFYCLUSTERSETTING global privilege came from product, correct? @thtruo perhaps you can speak to the desired behavior, and provide an answer to @data-matt's feedback in the initial issue description. Also, CCing @abhinavg6.

**Comment by Santamaura** [comment]
> It seems like the decision to enlarge the scope of the MODIFYCLUSTERSETTING global privilege came from product, correct?

This is correct

**Comment by data-matt** [comment]
Could we have 2 roles?

MODIFYSQLCLUSTERSETTING and MODIFYCLUSTERSETTING.

Then on upgrade, migrate the existing MODIFYCLUSTERSETTING to MODIFYSQLCLUSTERSETTING? 

Let me know if its a big ask.

**Comment by abhinavg6** [comment]
Is this isolation issue related to SQL cluster settings only? Could there be another bucket where someone comes up tomorrow and asks for a specific change? Is there a user persona related bucketing when it comes to cluster settings?

**Comment by data-matt** [comment]
Is this isolation issue related to SQL cluster settings only? Yes  Could
there be another bucket where someone comes up tomorrow and asks for a
specific change? Not from my perspective.
Is there a user persona related bucketing when it comes to cluster
settings? Yes, application users.

**Comment by thtruo** [comment]
To shed some some light on the decision made last cycle, the motivation was to provide non-admins the ability to manage _all_ cluster settings. It was a miss on my part for not catching the existing behavior for the `MODIFYCLUSTERSETTING` role option, which only limits users to `sql.defaults.*`, so I can see how this throws a wrench for customers who are used to the original behavior and not the new one.

> Once upgraded to 22.2, there will be users that have access to all cluster settings and therefore able to alter security settings such as the hba.conf.

@data-matt I'm curious, who is responsible for managing non-`sql.defaults.*` cluster settings? AFAICT only admin SQL users can modify those. But since we've been talking about the theme of not gating functionality behind ADMIN anymore, in this case wouldn't the customer want the ability to grant non-admin SQL users the privilege to modify these (non-`sql.defaults.*`) cluster settings anyway? Let me know if I'm off the mark here. 

I also want to ack that if there was a need for this, we should have introduced a differently named system privilege to avoid confusion.

--
Here are a few ideas on how to move forward - I'm open to other suggestions as well as I'm sure others might have alternative ideas

**Short-term**
* Could we ask the customer to use the `MODIFYCLUSTERSETTING` role option instead of the system privilege in the meantime? This is a short term workaround while the customer is on 22.2

**Medium/Long-term** (for 23.1+)
* We could consider introducing additional cluster setting system privileges, similar to what you were suggesting @data-matt 
* Or, an alternative approach is to modify the required privileges needed to have a user use `ALTER ROLE ALL SET {sessionvar} = {val}` instead of setting `sql.defaults.*` cluster settings as recommended by ([docs](https://www.cockroachlabs.com/docs/stable/cluster-settings.html)). For example, we could introduce a system privilege called `SETSESSION` (not the final name) to gate the ability for a user to do this. 

Thoughts @data-matt?

Let me know @abhinavg6 @rafiss or @Santamaura if you have major concerns/objections with any of these ideas.

**Comment by rafiss** [comment]
I like the idea of steering them towards using `ALTER ROLE ALL SET {sessionvar} = {val}` (if they agree that it works for their use case). One thing to note: using that command currently requires the CREATEROLE **role option** (not global privilege); this privilege check is the same as the one Postgres makes. A global privilege could be added for that too, if we want to depart from strictly matching Postgres.

**Comment by data-matt** [comment]
@thtruo 

**Short term**

I'm surprised both behaviours for MODIFYCLUSTERSETTING exist in 22.2, can we verify thats a thing? Definitely gets us more time. @rafiss  / @Santamaura 

**Long term**
 My assumption is that it won't be granted out, so users won't be able to ALTER ROLE. Checking with the folks.

> We could consider introducing additional cluster setting system privileges, similar to what you were suggesting @data-matt

This would be much appreciated. Lets see.

**Comment by Santamaura** [comment]
> Could we ask the customer to use the MODIFYCLUSTERSETTING role option instead of the system privilege in the meantime? This is a short term workaround while the customer is on 22.2

Don't we fallback to checking role options if the user doesn't have the system privilege? See: https://github.com/cockroachdb/cockroach/blob/15f51905b33ce68782c919ae03195d27dbe0d01a/pkg/sql/set_cluster_setting.go#L66

Not just in this case but I think in general we do this for system privileges.

**Comment by data-matt** [comment]
@thtruo , its not suitable for users to use ALTER ROLE, as they need CREATEROLE privilege.

Additionally @Santamaura has given me the impression we don't have both behaviours for MODIFYCLUSTERSETTING in 22.2+

Therefore I have to ask for the option suggested previously:
> We could consider introducing additional cluster setting system privileges, similar to what you were suggesting @data-matt

**Comment by data-matt** [comment]
@thtruo , will an additional role be possible?

**Comment by thtruo** [comment]
Synced internally. We'll proceed with introducing an additional system privilege called `MODIFYSQLCLUSTERSETTING` which, when granted to a user, will only provide them permission to edit `sql.defaults.*` cluster settings (we'll backport to the next patch version of 22.2)

We will *not* be introducing an equivalent role option, as role options are deprecated. Customers should utilize the system privilege.

**Comment by thtruo** [comment]
Quick update after syncing with @rafiss @Santamaura @dikshant: 

- The new `MODIFYSQLCLUSTERSETTING` will be included in 23.1
- However, there is a risk with backporting this system privilege in 22.2.x as @rafiss raised. While the blast radius is small (errors may occur during a rolling upgrade process if the customer explicitly grants this new privilege), the potential impact is quite severe preventing SQL user logins 
- So we are using an alternative solution. We will instead add a new cluster setting in 22.2.x for an administrator to configure the behavior of `MODIFYCLUSTERSETTING` - by default it will behave as is in 22.2 where any privileged user can modify all cluster settings, or it can be set to allow the privileged user to only modify `sql.defaults.*` cluster settings. That change is tracked in #98155 (thanks @Santamaura) 

FYI @data-matt this is the workaround that will be made available in an upcoming patch release of 22.2
